chain/chain/src/chain.rs

```diff
@@ -2672,8 +2677,10 @@ impl Chain {
             for receipt_proof in receipt_proofs.iter() {
                 let ReceiptProof(receipts, shard_proof) = receipt_proof;
                 let ShardProof { from_shard_id, to_shard_id: _, proof } = shard_proof;
-                let receipts_hash = CryptoHash::hash_borsh(ReceiptList(shard_id, receipts));
-                let from_shard_index = prev_shard_layout.get_shard_index(*from_shard_id)?;
+                let receipts_hash = CryptoHash::hash_borsh(ReceiptList(target_shard_id, receipts));
+                let shard_layout =
+                    self.epoch_manager.get_shard_layout(block.header().epoch_id())?;
+                let from_shard_index = shard_layout.get_shard_index(*from_shard_id)?;
```

Comment:
I'm tempted to implement a debug assert feature where we add an ephemeral `version` to shard layout and shard id and we assert that shard id can only be used with shard layout with same version. It would require a lot of refactoring though :(